### DISCUSSION TITLE
JSX Feature: Optional Curly Braces for Attributes

### DISCUSSION THREAD

**Discussion by lxsmnsyc** [solidjs-community/solid-dsl#7] [Proposals]
This feature would allow JSX to allow optional usage of curly braces when providing JS expressions for attributes.

```html
<button onClick=increment>Increment</button>
<Button value=3.14159 />
```

## Considered

The feature would only apply to the following expressions:

- [`ArrayLiteral`](https://tc39.es/ecma262/#prod-ArrayLiteral)

```html
<For each=[1, 2, 3] />
```

- [`Identifier`](https://tc39.es/ecma262/#prod-Identifier)
  - [`NullLiteral`](https://tc39.es/ecma262/#prod-NullLiteral)
  - [`BooleanLiteral`](https://tc39.es/ecma262/#prod-BooleanLiteral)

```html
<button onClick=increment>Increment</button>
<button disabled=true>Example</button>
<Input value=null />
<Input value=undefined />
```

- [`StringLiteral`](https://tc39.es/ecma262/#prod-StringLiteral) 

```html
<a href="mailto:">Mail</a>
<a href='mailto:'>Mail</a>
```

- [`NumericLiteral`](https://tc39.es/ecma262/#prod-NumericLiteral)

```html
<Number value=3.14159 />
<Number value=0xDEADBEEF />
<Number value=0x10010110 />
```

- [`RegularExpressionLiteral`](https://tc39.es/ecma262/#prod-RegularExpressionLiteral)

```html
<InputMask pattern=/^\d+$/ />
```

- [`TemplateLiteral`](https://tc39.es/ecma262/#prod-TemplateLiteral)

```html
<Message value=`${greeting}, ${name}` />
```

- [`ParenthesizedExpression`](https://tc39.es/ecma262/#prod-ParenthesizedExpression)

```html
<button onClick=(() => increment(100))>Increment By 100</button>
```

## Not considered

Most of the rejected expressions relates to allowing whitespace for tokens that do not have opening/closing symbols (which allows definition of token boundaries).

- `FunctionExpression`, `AsyncFunctionExpression`, `GeneratorExpression`, `AsyncGeneratorExpression`, `ClassExpression` - The syntax has no initial opening/closing tokens that would allow proper identification. Would require curly braces or parentheses to work (see ParenthesizedExpression)
- `ArrowFunction` - allows optional whitespace, doesn't have initial opening/closing tokens for proper identification. Requires curly braces or parentheses.
- `MemberExpression`, `ImportMeta`, `OptionalChain` - allows optional whitespace between chains, doesn't have initial opening/closing tokens for proper identification. Requires curly braces or parentheses
- `ObjectLiteral` - conflict with curly braces.
- `NewExpression` - requires whitespace
- `CallExpression`, `ImportCall` - allows optional whitespace between `callee` and the parameter list.
- `UnaryExpression`, `AssignmentExpression`, `BinaryExpression`, `TernaryExpression` - allows optional whitespace

**Comment by DylanPiercey** [comment]
FYI Marko's parser handles this but in a way that supports the vast majority of JS/TS expressions as attributes, mostly the only cases that our parser would not treat as an expression is when it conflicts with the html syntax, eg `<tag x=a>b` for obvious reasons is not going to treat `a>b` as the expression.

It works by maintaining a stack for enclosed sections (think string literals, brackets, regexps) and then outside of any enclosed thing using a regex that will match TS/JS binary and unary operators (https://github.com/marko-js/htmljs-parser/blob/main/src/states/EXPRESSION.ts#L214-L235)

You can see the types of values that can be expressions in our test snapshot here: https://github.com/marko-js/htmljs-parser/blob/main/src/__tests__/fixtures/attr-operators-space-between/__snapshots__/attr-operators-space-between.expected.txt

Just thought it might be interesting!

**Reply by lxsmnsyc** [reply]
This is an interesting solution, unfortunately I only saw this as added complexity. It felt like `ParenthesizedExpression` is always the better approach to this, just my opinion.

**Reply by DylanPiercey** [reply]
Agreed it's a bit more complex, however on the parser level you're going to have to do parts of what I described anyways (even normal jsx has to keep track of enclosed expressions like strings). After that in a parser that understands the js it'd largely follow the same rules as parsing an object literals property value (slightly trickier for Marko since we don't parse the js in the first pass).

Personally I think the complexity in the parser for the rest of the stuff is worth it since it means that users don't need to worry about using only certain types of expressions as values, new users would find that confusing.

**Reply by lxsmnsyc** [reply]
I think the part where it becomes confusing is the consistency of expressions (because of the markup's limitation)

**Reply by DylanPiercey** [reply]
Could you elaborate?

**Reply by lxsmnsyc** [reply]
Since you mentioned that there are expressions in Marko that are supported but there are some that isn't, specifically those that conflict with  the markup. So I felt like some user would find it surprising.

**Reply by DylanPiercey** [reply]
In Marko there's not really much that can conflict. For HTML mode it's `>` and for concise mode its newlines and `--`.

**Reply by DylanPiercey** [reply]
I think when you see it syntax highlighted it's obvious why those two don't work.
